src/internal/document_api/mod.rs

```diff
@@ -795,18 +809,71 @@ impl TryFrom<&EncryptedDek> for EncryptedDekP {
         Ok(proto_edek)
     }
 }
-
+/// Result of recrypt encryption. Contains the encrypted DEKs and the encrypted (user) data.
+/// `RecryptionResult` is an intermediate value as it cannot be serialized to bytes directly.
+/// To serialize to bytes, first construct an `EncryptedDoc`
 #[derive(Debug, Clone)]
-pub struct EncryptionResult {
+struct RecryptionResult {
     edeks: Vec<EncryptedDek>,
     encrypted_data: AesEncryptedValue,
     encryption_errs: Vec<DocAccessEditErr>,
 }
 
+impl RecryptionResult {
+    fn into_edoc(self, header: DocumentHeader) -> EncryptedDoc {
+        EncryptedDoc {
+            value: self,
+            header,
+        }
+    }
+}
+
+/// An ironoxide encrypted document
+#[derive(Debug)]
+struct EncryptedDoc {
+    header: DocumentHeader,
+    value: RecryptionResult,
+}
+
+impl EncryptedDoc {
+    /// bytes of the encrypted data with the edoc header prepended
+    fn edoc_bytes(&self) -> Vec<u8> {
+        [
+            &self.header.pack().0[..],
+            &self.value.encrypted_data.bytes(),
+        ]
+        .concat()
+    }
+
+    /// associated EncryptedDeks for this EncryptedDoc
+    fn edek_vec(&self) -> Vec<EncryptedDek> {
+        self.value.edeks.clone()
+    }
+
+    /// binary blob for associated edeks, or error if encoding the edeks failed
+    fn edek_bytes(&self) -> Result<Vec<u8>, IronOxideErr> {
+        let proto_edek_vec_results: Result<Vec<_>, _> = self
+            .value
+            .edeks
+            .iter()
+            .map(|edek| edek.try_into())
```

Comment:
I'm surprised going _to_ proto can fail...

Reply:
It can only fail because an internal `EncryptedDek` could be holding a `TransformedValue` because of the `EncryptedValue` recrypt enum. This would be a bug in IronOxide if this ever happened. This is only happening on umanaged encrypt. Previously it was being masked there as that has to return a `Result` for other errors as well.

We can relax this if enum variants ever become "real" types.

Reply:
Ugh. That's too bad.